### PR TITLE
Ost prime unit tests

### LINKED ISSUE #72: OST Prime contract interact unit tests
Write mocha unit test for ost prime contract interact.

### PR DESCRIPTION
**Pull request by 0xsarvesh**
Fixes #72 
⚠️  Merge after #56 
OST Prime contract interact unit tests.

### CODE REVIEW

File: test/OSTPrime/_approveRawTx.js [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// ----------------------------------------------------------------------------
+//
+// http://www.simpletoken.org/
+//
+// ----------------------------------------------------------------------------
```
Comment (schemar) [inline test/OSTPrime/_approveRawTx.js]: Please remove all license headers.

File: test/OSTPrime/_approveRawTx.js [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+const chai = require('chai');
```
Comment (schemar) [inline test/OSTPrime/_approveRawTx.js]:
```suggestion
const { assert } = require('chai');
```

All files

File: test/OSTPrime/_approveRawTx.js [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+const Web3 = require('web3');
+const sinon = require('sinon');
+
+const assert = chai.assert;
```
Comment (schemar) [inline test/OSTPrime/_approveRawTx.js]:
```suggestion
```

File: test/OSTPrime/_approveRawTx.js [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+const OSTPrime = require('../../src/ContractInteract/OSTPrime');
+const AssertAsync = require('../../test_utils/AssertAsync');
+const SpyAssert = require('../../test_utils/SpyAssert');
+
+describe('OSTPrime.allowance()', () => {
```
Comment (schemar) [inline test/OSTPrime/_approveRawTx.js]:
```suggestion
describe('OSTPrime._approveRawTx()', () => {
```

File: test/OSTPrime/_approveRawTx.js [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+  it('should pass with correct params', async () => {
+    let mockTx = 'tx';
+    const spyApprove = sinon.replace(
+      ostPrime.contract.methods,
+      'approve',
+      sinon.fake.resolves(Promise.resolve(mockTx)),
+    );
+
+    const spyMethod = sinon.spy(ostPrime, '_approveRawTx');
```
Comment (schemar) [inline test/OSTPrime/_approveRawTx.js]:
```suggestion
    const spyApproveRawTx = sinon.spy(ostPrime, '_approveRawTx');
```
?

File: test/OSTPrime/_approveRawTx.js [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+    const spenderAddress = '0x0000000000000000000000000000000000000004';
+    const amount = '100';
+
+    const result = await ostPrime._approveRawTx(spenderAddress, amount);
```
Comment (schemar) [inline test/OSTPrime/_approveRawTx.js]:
```suggestion
    const tx = await ostPrime._approveRawTx(spenderAddress, amount);
```
or `resultingTransaction`

File: test/OSTPrime/_unwrapRawTx.js [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+  it('should pass with correct params', async () => {
+    let mockTx = 'tx';
+    const spyUnwrap = sinon.replace(
+      ostPrime.contract.methods,
+      'unwrap',
+      sinon.fake.resolves(Promise.resolve(mockTx)),
+    );
+
+    const spyMethod = sinon.spy(ostPrime, '_unwrapRawTx');
```
Comment (schemar) [inline test/OSTPrime/_unwrapRawTx.js]:
```suggestion
    const spyUnwrapRawTx = sinon.spy(ostPrime, '_unwrapRawTx');
```

File: test/OSTPrime/_wrapRawTx.js [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+  it('should pass with correct params', async () => {
+    let mockTx = 'tx';
+    const spyUnwrap = sinon.replace(
+      ostPrime.contract.methods,
+      'wrap',
+      sinon.fake.resolves(Promise.resolve(mockTx)),
+    );
+
+    const spyMethod = sinon.spy(ostPrime, '_wrapRawTx');
```
Comment (schemar) [inline test/OSTPrime/_wrapRawTx.js]:
```suggestion
    const spyWrapRawTx = sinon.spy(ostPrime, '_wrapRawTx');
```

File: test/OSTPrime/allowance.js [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+    let expectedAllowance = '100';
+    const spyAllowance = sinon.replace(
+      ostPrime.contract.methods,
+      'allowance',
+      sinon.fake.returns({
+        call: () => Promise.resolve(expectedAllowance),
+      }),
+    );
+
+    const spyMethod = sinon.spy(ostPrime, 'allowance');
```
Comment (schemar) [inline test/OSTPrime/allowance.js]: Here you would have a naming conflict. Maybe you can find better non-conflicting names.

File: test/OSTPrime/approve.js [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+    const txOptions = { from: '0x0000000000000000000000000000000000000004' };
+
+    const result = await ostPrime.approve(spenderAddress, amount, txOptions);
+
+    assert.isTrue(result, 'Approve should return true');
+
+    SpyAssert.assert(approveRawTxSpy, 1, [[spenderAddress, amount]]);
+
+    SpyAssert.assert(spySendTransaction, 1, [[mockTX, txOptions]]);
+    sinon.restore();
```
Comment (schemar) [inline test/OSTPrime/approve.js:45]: Could it lead to problems to add this to `afterEach`?

File: test/OSTPrime/_approveRawTx.js [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// ----------------------------------------------------------------------------
+//
+// http://www.simpletoken.org/
+//
+// ----------------------------------------------------------------------------
+
```
Comment (hobofan) [inline test/OSTPrime/_approveRawTx.js]: license 🗑

File: test/OSTPrime/balanceOf.js [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+
+    const spyBalanceOf = sinon.replace(
+      ostPrime.contract.methods,
+      'balanceOf',
+      sinon.fake.returns({
+        call: () => Promise.resolve(balance),
+      }),
+    );
+
+    const spyMethod = sinon.spy(ostPrime, 'balanceOf');
```
Comment (schemar) [inline test/OSTPrime/balanceOf.js]: Would be a naming conflict again 🤔

**Review by hobofan — CHANGES_REQUESTED** [review]
License headers need to be removed. Apart from that it looks good 👍

File: test/OSTPrime/isAmountApproved.js [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+  let ostPrime;
+
+  beforeEach(() => {
+    web3 = new Web3();
+    ostPrimeAddress = '0x0000000000000000000000000000000000000002';
+    ostPrime = new OSTPrime(web3, ostPrimeAddress);
+  });
+
+  it(
+    'should return true if account approval is equal to expected' + ' amount',
```
Comment (schemar) [inline test/OSTPrime/isAmountApproved.js]: line breaks on purpose?

File: test/OSTPrime/isAmountApproved.js [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+        ostPrime,
+        'allowance',
+        sinon.fake.resolves(100),
+      );
+
+      const ownerAddress = '0x0000000000000000000000000000000000000002';
+      const spenderAddress = '0x0000000000000000000000000000000000000003';
+      const amount = '100';
+
+      const result = await ostPrime.isAmountApproved(
```
Comment (schemar) [inline test/OSTPrime/isAmountApproved.js]:
```suggestion
      const isApproved = await ostPrime.isAmountApproved(
```
?

File: test/OSTPrime/isAmountApproved.js [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+      assert.isTrue(result, 'Approve should return true');
+
+      SpyAssert.assert(allowanceSpy, 1, [[ownerAddress, spenderAddress]]);
+    },
+  );
+
+  it(
+    'should return true if account approval is more than expected' + ' amount',
+    async () => {
+      const allowanceSpy = sinon.replace(
```
Comment (schemar) [inline test/OSTPrime/isAmountApproved.js]: I actually like `...Spy`. Maybe rename the other `spy...`?
Anyways, should be consistent across tests if possible.

**Review by schemar — CHANGES_REQUESTED** [review]
Niiiice 🎉 

Again only style and typo comments 👍

File: test/OSTPrime/isAmountApproved.js (hunk 13 above)
Comment (0xsarvesh) [reply test/OSTPrime/isAmountApproved.js]: This is stupid linter I guess.

File: test/OSTPrime/approve.js (hunk 10 above)
Comment (0xsarvesh) [reply test/OSTPrime/approve.js:45]: It's not used for all tests, So I thought to not keep in afterEach.

**Review by hobofan — APPROVED** [review]

**Comment by schemar** [comment]
Please add `use strict`; to the beginning of all files. Sorry I forgot 👶

**Review by schemar — APPROVED** [review]
:shipit:

**Review by deepesh-kn — APPROVED** [review]
LGTM

**Pull request merged**